components/service/glean/src/main/java/mozilla/components/service/glean/storages/TimingDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,277 @@
+            val jsonObject: JSONObject
+            try {
+                jsonObject = JSONObject(json)
+            } catch (e: org.json.JSONException) {
+                return null
+            }
+
+            // Category can be empty so it may be possible to be a null value so try and allow this
+            // by using `orEmpty()` to fill in the value.  Other values should be present or else
+            // something is wrong and we should return null.
+            val category = jsonObject.tryGetString("category").orEmpty()
+            val name = jsonObject.tryGetString("name") ?: return null
+            val bucketCount = jsonObject.tryGetInt("bucketCount") ?: return null
+            // If 'range' isn't present, JSONException is thrown
+            val range = try {
+                jsonObject.getJSONArray("range").toList<Long>()
+            } catch (e: org.json.JSONException) {
+                return null
+            }
+            val histogramType = HistogramType.values()[jsonObject.tryGetInt("histogramType")
+                ?: return null]
+            // Attempt to parse the values map, if it fails then something is wrong and we need to
+            // return null.
+            val values = try {
+                val mapData = jsonObject.getJSONObject("values")
+                val valueMap: MutableMap<String, Long> = mutableMapOf()
+                mapData.keys().forEach { key ->
+                    valueMap[key] = mapData.tryGetLong(key) ?: 0L
+                }
+                valueMap
+            } catch (e: org.json.JSONException) {
+                // This should only occur if there isn't a key/value pair stored for "values"
+                return null
+            }
+            val sum = jsonObject.tryGetLong("sum") ?: return null
+            val timeUnit = TimeUnit.values()[jsonObject.tryGetInt("timeUnit")
+                ?: return null]
+
+            return TimingDistributionData(
+                category = category,
+                name = name,
+                bucketCount = bucketCount,
+                range = range,
+                histogramType = histogramType,
+                values = values,
+                sum = sum,
+                timeUnit = timeUnit
+            )
+        }
+    }
+
+    // This is a helper property to build the correct identifier for the metric and allow for
+    // blank categories
+    internal val identifier: String = if (category.isEmpty()) { name } else { "$category.$name" }
+
+    // This is a list of limits for the buckets.  Instantiated lazily to ensure that the range and
+    // bucket counts are set first.
+    internal val buckets: List<Long> by lazy { getBuckets() }
+
+    /**
+     * Accumulates a sample to the correct bucket
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun accumulate(sample: Long) {
+        for (i in buckets.indices) {
```

Comment:
I'm not sure you need the `buckets` list and to loop through it.  Couldn't you just do something like (pseudo code, because I don't know the specifics of Kotlin math)...

```
val index = Math.floor(Math.log(sample)) as Int
values["$index"] += 1
```

I don't know if natural log or log base 10 is more what people want or are used to.  But my point is just that there should be a mathematical function from value to bucket.  And this also creates the "infinite buckets" thing that was talked about on slack.

Reply:
I much prefer this method.  It's going to have overhead Math overhead for every accumulation but probably no more than the linear search.  This may also help with another issue I've been fighting.

Reply:
As much as I like this approach, in the interest of MVP I'm going to stick with the search approach rather than the calculation approach.  I was struggling to try and account for the min/max range and bucket counts effect on the bucket sizes, so it would be a good iteration when we add infinite buckets feature.